docs/strings.md

```diff
@@ -0,0 +1,464 @@
+	 * encoding) are copied into that buffer, or returns an immutable array
+	 * that may point to the internal storage for the string.
+	 *
+	 * If the caller first calls `fastest_unicode_encoding` and uses the result
+	 * as the generic parameter then it will return an internal buffer for
+	 * immutable strings.  This is intended to be used to build fast iterators.
+	 */
+	func copy_or_view_data[StringEncoding Enc](self: readonly & String,
+	                                           range: Range,
+	                                           buffer: Optional[Array[Enc.CharType] & mut])
+	     : Array[Enc.CharType] & (mut | imm);
+
+	/******************************************************************************
+	 * Operations on mutable strings
+	 *****************************************************************************/
+
+	/**
+	 * Updates a specific character. 
+	 */
+	func update(self: (mut | iso) & String, index: U64, c: Unichar);
+
+	/**
+	 * Replace the specified range in a string with another string.
+	 */
+	func update_range[String Str](self: (mut | iso) & String,
+	                              range: Range,
+	                              newString: Str & imm);
+}
+```
+
+For discussion, the above proposal is somewhat simplified:
+The `update` operation should take a `StringView`, a non-owning view of a substring.
+The `String` interface should also include an operation for providing a `StringView` on a specific range.
+The Verona region system makes tracking ownership of string views relatively simple:
+
+ - By default, all mutable strings are `iso`.
+ - Mutable string views are `mut`, in the same region as their owning string.
+ - Immutable strings are globally immutable (`imm`).
+ - Immutable string views are immutable.
+
+This makes it easy to ensure that the lifetime of a string view is bound to the lifetime of the underlying string.
+Once stack types are better defined, it should also be possible for string views to be stack types associated with the region of the string.
+
+The minimal standard library should also include:
+
+ - Concrete instantiations of the `String` interface, storing ASCII, UTF-8, and UTF-16.
+ - Operations for finding substrings and extracting substrings as string views.
+ - Operations for constructing the most size or speed-efficient concrete string type from a string literal or string view.
+ - Operations for computing efficient hashes on strings.
+
+Requirements on the language
+----------------------------
+
+Supporting a rich Verona string library imposes a few requirements on the language.
+Most of these are already in progress but are captured here for completeness.
+
+# Literal syntax
+
+String literals in the source code should be an opaque immutable type that implements an interface that allows bytes to be iterated.
+We then need lightweight syntax for passing an instance of this type to the constructor for a concrete string type.
```

Comment:
I think we want the standard library to do so.  If source code is always UTF-8, it can expand to UTF-32 and then pick the right encoding (or, at least, scan the input and calculate the best size for each encoding).  We actually do this in the compiler for ObjC: https://github.com/llvm/llvm-project/blob/31b3f32104db1293aec74ce865f7891ec8019090/clang/lib/CodeGen/CGObjCGNU.cpp#L1046

I want to move all of this logic to the standard library so that programmers can easily substitute their own policy.